noxfile.py

```diff
@@ -21,51 +30,138 @@
+CI: Final[bool] = "CI" in os.environ
 
 # used to reset cached coverage data once for the first test run only
 reset_coverage = True
 
+if TYPE_CHECKING:
+    ExecutionGroupType = object
+else:
+    ExecutionGroupType = Dict[str, Any]
+
+
+@dataclasses.dataclass
+class ExecutionGroup(ExecutionGroupType):
+    sessions: Tuple[str, ...] = ()
+    python: Optional[str] = MIN_PYTHON
+    project: bool = True
+    extras: Tuple[str, ...] = ()
+    groups: Tuple[str, ...] = ()
+    dependencies: Tuple[str, ...] = ()
+    experimental: bool = False
+    pyright_paths: Tuple[str, ...] = ()
+
+    def __post_init__(self) -> None:
+        if self.pyright_paths and "pyright" not in self.sessions:
+            raise TypeError("pyright_paths can only be set if pyright is in sessions")
+        if self.python in EXPERIMENTAL_PYTHON_VERSIONS:
+            self.experimental = True
+        for key in self.__dataclass_fields__.keys():
+            self[key] = getattr(self, key)  # type: ignore
+
+
+EXECUTION_GROUPS: List[ExecutionGroup] = [
+    ## pyright
+    *(
+        ExecutionGroup(
+            sessions=("pyright",),
+            python=python,
+            pyright_paths=("disnake", "tests", "examples", "noxfile.py", "setup.py"),
+            project=True,
+            # FIXME: orjson doesn't yet support python 3.14, remove once we migrate to uv and have version-specific locks
+            extras=("speed", "voice") if python not in EXPERIMENTAL_PYTHON_VERSIONS else ("voice",),
+            groups=("test", "nox"),
+            dependencies=("setuptools", "pytz", "requests"),  # needed for type checking
+        )
+        for python in ALL_PYTHONS
+    ),
+    # docs and pyright
+    ExecutionGroup(
+        sessions=("docs", "pyright"),
+        pyright_paths=("docs",),
+        extras=("docs",),
+    ),
+    # codemodding and pyright
+    ExecutionGroup(
+        sessions=("codemod", "autotyping", "pyright"),
+        pyright_paths=("scripts",),
+        groups=("codemod",),
+    ),
+    # the other sessions, they don't need pyright, but they need to run
+    ExecutionGroup(
+        sessions=("lint", "slotscheck", "check-manifest"),
+        groups=("tools",),
+    ),
+    ## testing
+    *(
+        ExecutionGroup(
+            sessions=("test",),
+            python=python,
+            groups=("test",),
+        )
+        for python in ALL_PYTHONS
+    ),
+    # coverage
+    ExecutionGroup(
+        sessions=("coverage",),
+        project=False,
+        groups=("test",),
+    ),
+]
+
 
-def install_deps(
-    session: nox.Session,
-    *,
-    extras: Sequence[str] = (),
-    groups: Sequence[str] = (),
-    dependencies: Sequence[str] = (),
-    project: bool = True,
-) -> None:
+def get_groups_for_session(name: str) -> List[ExecutionGroup]:
+    return [g for g in EXECUTION_GROUPS if name in g.sessions]
+
+
+def get_version_for_session(name: str) -> str:
+    versions = dict.fromkeys(g.python for g in get_groups_for_session(name) if g.python)
```

Comment:
```suggestion
    versions = set(g.python for g in get_groups_for_session(name))
```
now that order doesn't matter, this can be a simple set